common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,379 @@
+                    public void onCancel() {
+                        request.cancel();
+                    }
+                });
+                pinDialog.show();
+                //Set currentDialog to pinDialog
+                synchronized (smartcardDialogLock) {
+                    mCurrentDialog = pinDialog;
+                }
+            }
+        };
+    }
+
+    //Upon a positive button click in the PIN prompt, verify the provided PIN and handle the results.
+    public SmartcardPinDialog.PositiveButtonListener getSmartcardPinDialogPositiveButtonListener(final YubiKitCertDetails certDetails, final SmartcardPinDialog pinDialog,final ClientCertRequest request) {
+        final String methodTag = TAG + ":getSmartcardPinDialogPositiveButtonListener";
+
+        return new SmartcardPinDialog.PositiveButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick(final String pin) {
+                synchronized (deviceLock) {
+                    // Need to request a SmartCardConnection in order to access certs on YubiKey.
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                        @Override
+                        public void invoke(Result<UsbSmartCardConnection, IOException> value) {
+                            try {
+                                final SmartCardConnection c = value.getValue();
+                                final PivSession piv = new PivSession(c);
+                                //Call YubiKit method to verify PIN.
+                                try {
+                                    piv.verifyPin(pin.toCharArray());
+                                    //If pin is successfully verified, we will get the certificate and perform the rest of the logic for authentication.
+                                    X509Certificate cert = piv.getCertificate(certDetails.getSlot());
+                                    //TODO: do a confirmation that the cert retrieved above is the same cert that was picked earlier by comparing with YubiKitCertDetails
+                                    //TODO: Complete authentication using cert and YubiKit sdk.
+                                    //NOTE: below is for testing. This would get replaced by the logic for actual authentication.
+                                    //START of code for testing
+                                    Logger.infoPII(methodTag,  cert.getSubjectDN().getName());
+                                    mActivity.runOnUiThread(new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            Toast.makeText(mActivity.getApplicationContext(), "Success!", Toast.LENGTH_LONG).show();
```

Comment:
Is this going to be shipped? or is this for testing only?

Reply:
Testing - I marked the parts for testing from 322 - 337

Reply:
Let's add a todo here to make sure we remove it before we ship.